### ISSUE TITLE
Function App connection to KeyVault through VNET

### ISSUE THREAD

**Issue by JV-HCA** [Azure/Azure-Functions#2023]
Having an issue with an Azure Function App not being able to Keyvault, Function-App configured with VNET and KeyVault configured with same VNET and Firewall. All built with Terraform

Function app configuration was complaining of not being able to find the KeyVault. 

![image](https://user-images.githubusercontent.com/69352089/131019519-91590fc2-1e44-4767-b407-2176c23e59d2.png)

Ran the troubleshooting to review the vnet config and most everything came back clean. 

![image](https://user-images.githubusercontent.com/69352089/131018982-832ccb12-4f50-4b9c-8b1a-c366ceb3ebd7.png)

Still no dice on the keyvault connection. Was reading about adding the "Custom Domain IP" for the function app to the keyvault firewall. I added that and it worked.
 (where y.y.y.y/32 is the function app ip)

```
network_acls {
    bypass                     = "AzureServices"
    default_action             = "Deny"
    ip_rules                   = ["x.x.x.x/x", "x.x.x.x/x", "x.x.x.x/x","y.y.y.y/32"]
    virtual_network_subnet_ids = [data.azurerm_subnet.subnet.id]
  }
```

I got to thinking and talking with my team, it should work without it because of the vnet integration and wanted to troubleshoot more.
Removed the Function App IP from the firewall and It's still working?! 
```
network_acls {
    bypass                     = "AzureServices"
    default_action             = "Deny"
    ip_rules                   = ["x.x.x.x/x", "x.x.x.x/x", "x.x.x.x/x"]
    virtual_network_subnet_ids = [data.azurerm_subnet.subnet.id]
  }
```
![image](https://user-images.githubusercontent.com/69352089/131020387-503ca857-79bb-487b-83cf-80b380681ccc.png)

Its almost as if you had to give it a nudge and it worked itself out. It's really strange. 

Anyone else seen similar behavior? If so, is my fix short lived and will it break again soon?

_Originally posted by @JV-HCA in https://github.com/Azure/Azure-Functions/issues/1291#issuecomment-906663857_

**Comment by pragnagopa** [comment]
@sidkri / @pgombar - can you please follow up on custom domain requirement?

**Comment by Monishguru** [comment]
> 2021-08-27T12:41:21.023 Id=cf7d087b-385d-4c0b-8e2c-3ddc20721ffe, Duration=7054ms
> 
> Error is not currently occurring. If you read through my issue I noted that when I set it up initially the function app said it couldn't find the KeyVault. After I added the "'custom domain" IP to the KeyVault firewall it was able to connect. to verify that is what fixed it I removed the IP from the firewall and it is still working. Based on the configuration, it should have worked without that in the first place because of the way I have vnet configured on the FunctionApp and the KeyVault. Its almost like I when I added the custom domain it established the route that should have existed and then when I removed it kept the route in place.
> 
> My biggest concern now is if that connection is stable without the 'custom domain' IP being in place in the firewall, and if at any point it will revert and that connection drop again.

Exactly, experiencing the same here...Maybe after removing the Custom Ip and doing a terraform apply, and refreshing the Function App once removes the access to Key Vault.

**Comment by janitorr** [comment]
Hello, I know this has been a bit quiet lately but this has recently gotten our application crash twice in production and we were able to fix it by redeploying.

What is the current recommendation to mitigate the issue, I was planning to inject the configs in the app in deploy time but are there other more recommended alternatives to this?

**Comment by Maximum50** [comment]
@JV-HCA , I'm also getting the exact same error, and was trying to figure out how did this go wrong.
Even when we have configured all vnet integration and key-vault references correctly.

Feels so sad. But your post is really helpful. I'll try out your solution and post here if I manage to fix temporarily.

**Comment by Maximum50** [comment]
ROFL,  I tried the work-around fix, and it did not work for me straight away..

So I removed and added the managed identity again, then restarted function app few times etc.
Removed the IP and added it again etc.
I have no idea what azure is doing, but suddenly started working

Logic just went out the window !

![image](https://user-images.githubusercontent.com/22293047/201331480-47c07214-82f7-4266-b253-8f82d39f5a24.png)

**Comment by davidzenisu** [comment]
Same user-experience 20 months later. Connection reported error; turned off firewall of the key vault completely then back on again and suddenly it just kept working. Really frustrating fiddling with these setups in general...

**Comment by lena-osterhus** [comment]
We just had the same issue in production where Key Vault Resources suddenly could not be resolved. Work around with temporarily adding Azure Function app [outbound IP addresses](https://learn.microsoft.com/en-us/azure/azure-functions/ip-addresses?tabs=azurecli#find-outbound-ip-addresses) to KeyVault firewall worked.
